### DISCUSSION TITLE
GraphQL over Server-Sent Events Protocol Server

### DISCUSSION THREAD

**Discussion by rishabhjha708** [Netflix/dgs-framework#1124] [Q&A]
Can we make GraphQL over Server-Sent Events Protocol server from DGS framework. I checked the doc, only found the client support for that.
@paulbakker  can you please help here, any way to achieve that ?

**Comment by Ancient-Dragon** [comment]
See the PR here: https://github.com/Netflix/dgs-framework/pull/1122 does this work for you?

**Reply by rishabhjha708** [reply]
Hi @Ancient-Dragon I wanted to understand, is this for consuming subscription or creating subscription, can we create subscription without websocket over SSE ?

**Reply by Ancient-Dragon** [reply]
That PR allows you to host a graphql server which supports SSE subscriptions (i.e. a consumer can subscribe to your server for updates via SSE). 

There is an SSE client in the DGS framework which allows you to query other services via SSE but it only uses GET requests rather than using POST.

**Reply by rishabh-jha708** [reply]
@Ancient-Dragon I was checking your [Example backend project](https://github.com/Ancient-Dragon/dgs-sse-subscriptions-backend-example), I ran the project successfully by changing gradle import of autoConfigure of SSE "com.netflix.graphql.dgs:graphql-dgs-subscriptions-sse-autoconfigure:5.0.5" as Snapshot version was unavailable
but when I tried to run the query using altair, it gives me following error. Can you guide me on this.

<img width="663" alt="image" src="https://user-images.githubusercontent.com/101696098/177193419-465f79ce-76f4-420a-9413-94b9e1f55c04.png">

<img width="1398" alt="image" src="https://user-images.githubusercontent.com/101696098/177193240-b5eb1b83-f070-4698-b9bd-40c2ff30203c.png">

**Reply by Ancient-Dragon** [reply]
Ah yes, that version of altair is old / doesn't seem to work. If you get the latest version of altair as an extension etc you should be able to see this screen when configuring the subscriptions url:
![Screenshot 2022-07-04 174049](https://user-images.githubusercontent.com/22601875/177193719-853a421f-45a6-4384-885c-85dbb16d50e4.png)
 As you can see I have configured subscriptions to use SSE

**Reply by rishabh-jha708** [reply]
@Ancient-Dragon Yes using latest chrome extension of altair I was able to get that option.
But not receiving any logs in server

<img width="1392" alt="image" src="https://user-images.githubusercontent.com/101696098/177194932-82a42e1e-9fdb-43eb-8e8a-94e675e7bf7d.png">

<img width="1257" alt="image" src="https://user-images.githubusercontent.com/101696098/177195000-9b37a34c-74e0-43d2-8e6b-3753f68acb80.png">

**Reply by Ancient-Dragon** [reply]
Did you configure your chrome to allow self signed certificates on localhost?

**Reply by rishabh-jha708** [reply]
@Ancient-Dragon  Yes, I did that I am able to run the project.

So, as per my understanding, We can develop graphql subscription server over SSE just we have to add graphql-dgs-subscriptions-sse-autoconfigure as dependency for SSE. correct me if I am wrong ?

**Reply by Ancient-Dragon** [reply]
That is correct :)

**Reply by rishabh-jha708** [reply]
Hi @Ancient-Dragon , 1 doubt the PR which you shown it added support for post request on subscription over SSE, right
<img width="459" alt="image" src="https://user-images.githubusercontent.com/101696098/177204484-d837c01f-5a94-42ff-a423-3ba938507f1d.png">
But how it is automatically calling PostMapping request ? I checked in your example project but did not find any explicit declaration.

**Reply by Ancient-Dragon** [reply]
What do you mean it is calling post mapping request, could you explain what you are trying to achieve with this discussion?

**Reply by rishabh-jha708** [reply]
Just wanted to understand about the PR that you mentioned. Rest I got the clarity how we can create subscription over SSE.

**Reply by Ancient-Dragon** [reply]
So the PR was changing a class that is a rest controller which just had a request mapping which would take all methods but required a parameter in the url called query (which is what you should do if you want to do graphql over GET). However if you are using post (as most libraries do) then the body is where the query is. Therefore I created a new PostMapping and altered the current code to be a GetMapping. Behind the scenes this allows spring to automatically handle the GET and POST requests accordingly to their respective methods. Hopefully I've explained that ok. Using the auto-configration library just creates a bean of this RestController.

**Reply by rishabh-jha708** [reply]
Got it Thanks @Ancient-Dragon
